All properties being converted to strings.

In DriverDataSource line 49 all properties are getting converted to strings:

      for (var entry : properties.entrySet()) {
         driverProperties.setProperty(entry.getKey().toString(), entry.getValue().toString());
      }

This is an issue for the Snowflake driver because the only successful way to connect with an encrypted private key is to add the property "privateKey" with a value of an object of type java.security.PrivateKey. When that object gets converted to a string the connection fails with the following error:
Caused by: net.snowflake.client.jdbc.SnowflakeSQLLoggedException: Invalid parameter value type: java.lang.String, expected type: java.security.PrivateKey.



Changing the line to the following results in a successful connection:
      for (var entry : properties.entrySet()) {
         driverProperties.put(entry.getKey(), entry.getValue());
      }

Can you advise? I hate to have to override the DriverDataSource class locally.

--- a/src/main/java/com/zaxxer/hikari/util/DriverDataSource.java
+++ b/src/main/java/com/zaxxer/hikari/util/DriverDataSource.java
@@ -45,9 +45,7 @@ public final class DriverDataSource implements DataSource
       this.jdbcUrl = jdbcUrl;
       this.driverProperties = new Properties();
 
-      for (var entry : properties.entrySet()) {
-         driverProperties.setProperty(entry.getKey().toString(), entry.getValue().toString());
-      }
+      driverProperties.putAll(properties);
 
       if (username != null) {
          driverProperties.put(USER, driverProperties.getProperty(USER, username));

--- a/src/test/java/com/zaxxer/hikari/util/DriverDataSourceTest.java
+++ b/src/test/java/com/zaxxer/hikari/util/DriverDataSourceTest.java
@@ -18,6 +18,8 @@ package com.zaxxer.hikari.util;
 
 import org.junit.Test;
 
+import java.lang.reflect.Field;
+import java.time.Duration;
 import java.util.Arrays;
 import java.util.List;
 import java.util.Properties;
@@ -25,6 +27,18 @@ import java.util.Properties;
 import static org.junit.Assert.*;
 
 public class DriverDataSourceTest {
+
+   @Test
+   public void testDriverProperties() throws Exception {
+      Properties properties = new Properties();
+      Duration timeout = Duration.ofSeconds(60);
+      properties.put("timeout", timeout);
+      var driverDataSource = new DriverDataSource("jdbc:h2:mem:test;DB_CLOSE_DELAY=-1", null, properties, "", "");
+      Field field = DriverDataSource.class.getDeclaredField("driverProperties");
+      field.setAccessible(true);
+      Properties driverProperties = (Properties) field.get(driverDataSource);
+      assertEquals(timeout, driverProperties.get("timeout"));
+   }
 
    @Test
    public void testJdbcUrlLogging() {

